feat: Add Library Collections REST endpoints [FC-0062]

yusuf-musleh opened the pull request:
## Description

This PR adds REST endpoints (list/create/update/delete) for Library Collections, in addition to emitting events when a library is created/updated. The deleted endpoint and emitting its event will be implemented in a later ticket, however the implementation skeleton is there.

## Supporting information

Related Tickets:
 - https://github.com/openedx/modular-learning/issues/226

Depends on / blocked by:
 - https://github.com/openedx/openedx-events/pull/386
 - https://github.com/openedx/openedx-learning/pull/223

## Testing instructions

1. Go over new tests, make sure they cover most cases and make sure they are passing
1. For additional manual testing:
    1. Checkout this branch in your local dev environment
    1. Make sure you have https://github.com/openedx/openedx-events/pull/386 + https://github.com/openedx/openedx-learning/pull/223 installed and running locally 
    1. Make sure you already have a V2 library or created one if not, e.g `lib:SampleTaxonomyOrg1:AL1`
    1. Add some blocks to that library.
    1. In an external HTTP client (like curl/postman or similar) make the following requests (make sure you have the correct headers for authentication, you can copy them from the network tab in your browser), replace the library key in the URL with your actual library key:
        1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/ - Should return a list of all collections that belong to the library, initially it will be empty
        1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/ - Should return a specific collection (with key="COL1") that belongs to the library, initially this should return 404, not found
        1. POST http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/
            1. Payload: {"key": "COL1", "title":"Collection 1", "description": "Desc for Collection 1"} 
            1. This should return the newly created collection
        1. PATCH http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/
            1. Payload: {"title": "Collection 1 updated title"}
            1. This should return the updated collection
        1. POST http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/
            1. Payload: {"key": "COL2", "title":"Collection 2", "description": "Desc for Collection 2"} 
            1. Should return another newly created collection
        1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/ - This should now return the 2 collections created above
        1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/ - Should return the first collection we created
        1. DELETE http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/ - Should not delete and return a 405
       1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/blocks/
          1. Should return the list of blocks added to this library
          1. Note their usage key values for the steps below.
       1. PATCH http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/contents/
           1. Payload: `{"usage_keys": [ <list of usage keys from the library blocks> ]`
           1. Should return `{"count": <number of components in list>}`
           1. Check the logs and confirm that the event was emitted for each component, triggering an "upsert" into Meilisearch.
       1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/
           1. The `modified` timestamp should be updated.
       1. DELETE http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/contents/
           1. Payload: `{"usage_keys": [ <list of usage keys from the library blocks> ]`
           1. Should return `{"count": <number of components in list>}`
           1. Check the logs and confirm that the event was emitted for each component, triggering an "upsert" into Meilisearch.
       1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/
           1. The `modified` timestamp should be updated.
       1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/contents/ - will raise a 405, as an object's collections will be fetched from Meilisearch.

---
Private-Ref: [FAL-3783](https://tasks.opencraft.com/browse/FAL-3783)

yusuf-musleh committed:
feat: Add Library Collections REST endpoints

openedx-webhooks commented:
Thanks for the pull request, @yusuf-musleh!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.



#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

yusuf-musleh committed:
test: Add tests for Collections REST APIs

yusuf-musleh committed:
chore: Add missing __init__ files

pomegranited commented on openedx/core/djangoapps/content_libraries/collections/rest_api/v1/views.py:
```diff
@@ -0,0 +1,110 @@
+        # Check if user has permissions to view this collection by checking if
+        # user has permission to view the Content Library it belongs to
+        library_key = LibraryLocatorV2.from_string(lib_key_str)
+        api.require_permission_for_library_key(library_key, request.user, permissions.CAN_VIEW_THIS_CONTENT_LIBRARY)
```
@yusuf-musleh We need to make sure the collection is actually in the requested library.. what do you think of this approach?

```python
# TODO: Update api.require_permission_for_library_key to return the library object so we're not refetching
library_obj = api.require_permission_for_library_key(library_key, request.user, permissions.CAN_EDIT_THIS_CONTENT_LIBRARY)
collection = authoring_api.get_learning_package_collections(library_obj.learning_package_id).filter(id=pk).first()
```

yusuf-musleh replied:
@pomegranited Good catch! I like the approach you suggested, I placed it in a helper method for reuse in other endpoints. I also added more tests to check for this case. [743291c](https://github.com/openedx/edx-platform/pull/35321/commits/743291c8b4de541b890ea519180f602d19900d4b)

yusuf-musleh committed:
feat: Verify collection belongs to library

yusuf-musleh committed:
feat: Add events emitting for Collections

yusuf-musleh committed:
chore: fix pylint errors

rpenido commented on openedx/core/djangoapps/content_libraries/collections/rest_api/v1/views.py:
```diff
@@ -0,0 +1,181 @@
+        if library_obj.learning_package_id:
+            collection = authoring_api.get_learning_package_collections(
+                library_obj.learning_package_id
+            ).filter(id=collection_id).first()
```
After all, we ended up not using the `get_collection` API.
I think we should do one of the following:
1- Remove the `get_collection` function in the API to clean dead code
2- Update the `get_collection` function by adding a `learning_package_id` as a parameter

I'm happier with 2, as we move model logic from the view to the API. What do you think?

pomegranited replied:
@yusuf-musleh I agree with @rpenido , I think renaming `get_learning_package_collections` to `get_collections` makes sense -- there's no use case for fetching _all_ collections without limiting by learning package.

To keep from blocking this PR, I can do this as part of https://github.com/openedx/openedx-learning/pull/216 + https://github.com/openedx/edx-platform/pull/35384.

rpenido replied:
> there's no use case for fetching all collections without limiting by learning package

We are using it in the reindex command, but we can move it to the library loop without problems.

yusuf-musleh replied:
@rpenido Good suggestion! and thanks @pomegranited for implementing it as part of your PR!

rpenido commented on openedx/core/djangoapps/content_libraries/collections/rest_api/v1/views.py:
```diff
@@ -0,0 +1,181 @@
+                library_key, request.user, permissions.CAN_VIEW_THIS_CONTENT_LIBRARY
+            )
+        except api.ContentLibraryNotFound as exc:
+            raise Http404 from exc
```
We have a [`convert_exceptions`](https://github.com/openedx/edx-platform/blob/ea596d6a25e6608305be535e933146af015b4ec5/openedx/core/djangoapps/content_libraries/views.py#L129-L154) decorator in the `content_library` that we can move to a utils file and use in our functions here, so we don't need to re-raise our exceptions.

rpenido commented on openedx/core/djangoapps/content_libraries/collections/rest_api/v1/views.py:
```diff
@@ -0,0 +1,181 @@
+        except api.ContentLibraryNotFound as exc:
+            raise Http404 from exc
+
+        collection = None
```
If we use the `convert_exceptions` decorator and change the API, we can just make it throw if not found.

rpenido commented on openedx/core/djangoapps/content_libraries/collections/rest_api/v1/views.py:
```diff
@@ -0,0 +1,181 @@
+                library_key, request.user, permissions.CAN_EDIT_THIS_CONTENT_LIBRARY
+            )
+        except api.ContentLibraryNotFound as exc:
+            raise Http404 from exc
```
Same about `convert_exceptions`

rpenido commented on openedx/core/djangoapps/content_libraries/collections/rest_api/v1/views.py:
```diff
@@ -0,0 +1,181 @@
+        """
+        # TODO: Implement the deletion logic and emit event signal
+
+        return Response(None, status=HTTP_405_METHOD_NOT_ALLOWED)
```
~I prefer just to remove it. What do you think?~

Edit: solved [here](https://github.com/openedx/openedx-events/pull/386#issuecomment-2315818719)

rpenido commented on openedx/core/djangoapps/content_libraries/collections/rest_api/v1/views.py:
```diff
@@ -0,0 +1,181 @@
+        try:
+            library_obj = api.require_permission_for_library_key(library_key, user, permission)
+        except api.ContentLibraryNotFound as exc:
+            raise Http404 from exc
```
Ditto `convert_exceptions`

rpenido commented on openedx/core/djangoapps/content_libraries/collections/rest_api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,256 @@
+            URL_LIB_COLLECTION.format(lib_key=self.lib2.library_key, collection_id=self.col3.id)
+        )
+
+        assert resp.status_code == 405
```
~In case we remove the delete endpoint~

Edit: solved [here](https://github.com/openedx/openedx-events/pull/386#issuecomment-2315818719)

rpenido commented on openedx/core/djangoapps/content_libraries/collections/rest_api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,256 @@
+            format="json"
+        )
+
+        assert resp.status_code == 404
```
nit
```suggestion
        assert resp.status_code == 404
        
        # Update collection with invalid library_key provided, it should fail
        resp = self.client.patch(
            URL_LIB_COLLECTION.format(lib_key=123 collection_id=self.col3.id),
            patch_data,
            format="json"
        )

        assert resp.status_code == 404
```

rpenido commented on openedx/core/djangoapps/content_libraries/collections/handlers.py:
```diff
@@ -0,0 +1,57 @@
+        log.error("Received null or incorrect data for event")
+        return
+
+    log.info("Received Collection Deleted Signal")
```
~If we remove the `LIBRARY_COLLECTION_DELETED` event~

Edit: solved [here](https://github.com/openedx/openedx-events/pull/386#issuecomment-2315818719)

rpenido commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -323,13 +324,14 @@ def get_metadata(queryset, text_search=None):
 
 
-def require_permission_for_library_key(library_key, user, permission):
+def require_permission_for_library_key(library_key, user, permission) -> ContentLibrary:
```
:heart:

rpenido commented on docs/hooks/events.rst:
```diff
@@ -247,3 +247,15 @@ Content Authoring Events
+
+   * - `LIBRARY_COLLECTION_DELETED <https://github.com/openedx/openedx-events/blob/main/openedx_events/content_authoring/signals.py#L241>`_
+     - org.openedx.content_authoring.content.library.collection.deleted.v1
+     - 2024-08-23
```
~If we remove the `LIBRARY_COLLECTION_DELETED` event~

Edit: solved [here](https://github.com/openedx/openedx-events/pull/386#issuecomment-2315818719)

rpenido commented on openedx/core/djangoapps/content_libraries/collections/rest_api/v1/views.py:
```diff
@@ -0,0 +1,181 @@
+
+        # Check if user has permissions to create a collection in the Content Library
+        # by checking if user has permission to edit the Content Library
+        library_key = LibraryLocatorV2.from_string(lib_key_str)
```
If we pass an invalid lib_key_str here, this crashes

i.e `http://studio.local.edly.io:8001/api/libraries/v2/dsada/collections/`

pomegranited replied:
Worth handling as part of `convert_exceptions` I think?

yusuf-musleh replied:
Yes, having it handled as part of `convert_exceptions` would make sense since this issue also exists for all `/api/libraries/v2/[lib_key_str]` endpoints

rpenido commented on openedx/core/djangoapps/content_libraries/collections/rest_api/v1/views.py:
```diff
@@ -0,0 +1,181 @@
+
+        # Check if user has permissions to view collections by checking if user
+        # has permission to view the Content Library they belong to
+        library_key = LibraryLocatorV2.from_string(lib_key_str)
```
Same here

rpenido commented on openedx/core/djangoapps/content_libraries/collections/rest_api/v1/views.py:
```diff
@@ -0,0 +1,181 @@
+            raise Http404
+
+        pk = kwargs.pop("pk", None)
+        library_key = LibraryLocatorV2.from_string(lib_key_str)
```
And here

rpenido reviewed (changes requested):
Good work here @yusuf-musleh!

I found a small issue while passing invalid library keys and added some other comments.
Let me know what you think!

pomegranited commented on openedx/core/djangoapps/content_libraries/collections/rest_api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,256 @@
+
+from __future__ import annotations
+
+from openedx_learning.api.authoring_models import Collection
```
nit: Could you use the `openedx_learning.api.authoring` API instead of creating Collection objects directly?

yusuf-musleh committed:
refactor: Use convert_exceptions + update tests

yusuf-musleh commented:
@rpenido @pomegranited Thanks for the reviews/comments! I've addressed them all here: [5bf9422](https://github.com/openedx/edx-platform/pull/35321/commits/5bf94226fcbb778a28d63bb11ae10c43061f2a8b)

pomegranited commented on openedx/core/djangoapps/content_libraries/views.py:
```diff
@@ -120,40 +119,13 @@
 from openedx.core.djangoapps.xblock import api as xblock_api
 
 from .models import ContentLibrary, LtiGradedResource, LtiProfile
+from .utils import convert_exceptions
```
Do we actually need to move this function? I think it's ok to leave it in `content_libraries.views`.

yusuf-musleh replied:
Good point, I put it back and just imported it directly: [0bf0f78](https://github.com/openedx/edx-platform/pull/35321/commits/0bf0f78a33587c0b7d81762abbaf16782097b01b)

pomegranited commented on openedx/core/djangoapps/content_libraries/collections/rest_api/v1/views.py:
```diff
@@ -0,0 +1,177 @@
+"""
+Collections API Views
```
Suggestion: since there's nothing else under `content_libraries/collections` besides these views, could they go in `content_libraries/views.py` or `content_libraries/views_collections.py` instead?

yusuf-musleh replied:
I was wondering if file structure of `rest_api/v1/views.py` is the _standard_ way of structuring the REST endpoints or is that the older way to do it? I've seen some with and some without the above structure. Either way I think it makes sense to have them in their own file, whether its inside a `/collections` dir or in `content_libraries/views_collections.py` as you suggested.

yusuf-musleh replied:
I moved the collections view/tests out to the `content_libraries/`, ended up being way less boilerplate files/folders. [366a7e9](https://github.com/openedx/edx-platform/pull/35321/commits/366a7e9412a97034d0e473ee67f324cbcb97377c)

pomegranited commented on openedx/core/djangoapps/content_libraries/collections/rest_api/v1/views.py:
```diff
@@ -0,0 +1,177 @@
+                library_key=library_key,
+                collection_id=collection.id
+            )
+        )
```
Sorry about this, but I'm wondering if we need to move this functionality into `content_libraries.api`, so that anyone who creates, updates, or deletes a collection from the code gets the expected event emitted? The `content_libraries.api` already emits a bunch of signals, but the `content_libraries.views` don't, and it seems inconsistent to emit them here.

I did something like this for components, cf https://github.com/open-craft/edx-platform/pull/674/files#diff-a760c3e87f2f671ac2453abbbaf6f10ca052c7acf1c46c1d36f24646e1d52a00R1072

Same comment for `partial_update` below.

pomegranited replied:
But if we opt to send these `COLLECTION_CREATED` / `UPDATED` events from openedx-learning instead of here, then please ignore my previous comment.

yusuf-musleh replied:
Will hold on making changes to where the events are emitted from once the decision is made in the openedx-events PR, but I agree, having them in the `content_libraries.api` is much cleaner and it should reside there if we opted to keep the event emitting in edx-platform.

pomegranited commented on openedx/core/djangoapps/content_libraries/collections/handlers.py:
```diff
@@ -0,0 +1,57 @@
+"""
```
Since these handlers aren't actually doing anything yet, can this file be deleted?
We'll likely have the handlers live with `content/search/handlers.py`, since that's the app that cares about these events.

yusuf-musleh replied:
Good point, I mainly had it there to confirm that the signals were actually being sent, removed them here: [45cf886](https://github.com/openedx/edx-platform/pull/35321/commits/45cf88652a1991fc58b6917d87ddcab036c0f203)

pomegranited reviewed (changes requested):
Hi @yusuf-musleh , this is working great, but I've requested some structural changes to where the code lives.

yusuf-musleh committed:
refactor: Relocate convert_exceptions, remove utils

yusuf-musleh committed:
refactor: Remove Collection handlers skeleton code

yusuf-musleh committed:
refactor: Move Collections views/tests

bradenmacdonald commented:
Does this account for the new requirement that collections will need `slug`s, as discussed on the opaque-keys PR, or is that coming in a separate PR?

pomegranited reviewed (approved):
:+1: Thank you for those refactors @yusuf-musleh ! LGTM.

- [x] I tested this using the excellent PR test instructions
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] Includes documentation
- [x] User-facing strings are extracted for translation

pomegranited commented:
@bradenmacdonald 

> Does this account for the new requirement that collections will need slugs, as discussed on the opaque-keys PR, or is that coming in a separate PR?

No, that requirements change isn't handled here, but we can if needed.

This PR uses the `collection.id` to identify the collection in the REST API. Should it use this new `slug` instead? Or maybe the full LibraryCollectionLocator ?

rpenido reviewed (approved):
LGTM 👍
Thank you for your work, @yusuf-musleh !

- [x] I tested this using the instructions from the PR
- [x] I read through the code
- [x] ~I checked for accessibility issues~
- [x] Includes documentation

pomegranited committed:
feat: Add REST endpoints to update Components in a Collections (temp) (#674)

* chore: uses openedx-learning==0.11.3
* feat: add/remove components to/from a collection
  Sends a CONTENT_OBJECT_TAGS_CHANGED for each component added/removed.
* docs: Add warning about unstable REST APIs
* refactor: use oel_collections.get_collections as oel_collections.get_learning_package_collections has been removed.
* test: fixes flaky collection search test
* refactor: simplify the REST API params and validation

pomegranited committed:
refactor: pull view functionality into api

For the Collection "create" and "update" views, we call an authoring_api
method and emit an event on success. This change simplifies the view
code by moving these call to new content_libraries.api methods.

pomegranited commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1056,6 +1072,150 @@ def revert_changes(library_key):
     )
 
 
+def create_library_collection(
```
@bradenmacdonald  Could you have a look at these API methods to make sure you're ok with the format I chose?
Specifically is it ok to have the `content_library` as an optional parameter to avoid re-fetching that record?

bradenmacdonald replied:
Fine with me, but I would do it like this - with a comment to explain why, and requiring it as a kwarg so we can more easily change the preceding parameters.

```python
    description: str = "",
    created_by: int | None = None,
    *,
    # As an optimization, callers may pass in a pre-fetched ContentLibrary instance
    content_library: ContentLibrary | None = None,
) -> Collection:
```

Or maybe even move the `*` before `created_by` so it also has to be a kwarg.

pomegranited commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1056,6 +1072,150 @@ def revert_changes(library_key):
+    for usage_key in usage_keys:
+        CONTENT_OBJECT_TAGS_CHANGED.send_event(
+            content_object=ContentObjectData(object_id=usage_key),
+        )
```
@bradenmacdonald Since the `collection.modified` date changes here, I'm emitting a `LIBRARY_COLLECTION_UPDATED` event so the collection's metadata can be reindexed.

@rpenido [suggested on openedx-events#386](https://github.com/openedx/openedx-events/pull/386#issuecomment-2316518199) that, instead of emitting all of these `CONTENT_OBJECT_TAGS_CHANGED` events, we could give let the `LIBRARY_COLLECTION_UPDATED` handler reindex each of the collection's content objects. I could gate this using your suggested `reasons` field (or with just a boolean flag?) to see whether or not the entities have changed. Then we could leave `CONTENT_OBJECT_TAGS_CHANGED` unchanged.

Edit: I think I understand now why sending a `CONTENT_OBJECT_TAGS_CHANGED`-like event for each added/removed content object is better, even though it looks like we're "doing more things" -- Collections can have an unbounded number of objects, and so this way, we're only re-indexing the objects that are changing, not re-indexing every object in a Collection every time its contents changes.

But I wanted to make sure -- which way do you prefer?

bradenmacdonald replied:
I'd prefer to emit a `CONTENT_OBJECT_TAGS_CHANGED` (or `CONTENT_OBJECT_ASSOCIATIONS_CHANGED`) for each individual component. It's much more efficient to just push an updated document for that one component into the search index, rather than to re-index the entire collection. We want to avoid re-indexing a whole collection/library/course as much as possible. (I think the only time we currently _have_ to do it is on publishing or discarding changes, because we don't have a good way to know which individual components to reindex in that case.)

It's also totally fine to emit many small events. They'll get enqueued and processed in order, and it doesn't matter if it's slightly less efficient in some cases than one "big" event.

pomegranited commented on openedx/core/djangoapps/content_libraries/views_collections.py:
```diff
@@ -0,0 +1,184 @@
+        """
+        lib_key_str = kwargs.pop("lib_key_str")
+        library_key = LibraryLocatorV2.from_string(lib_key_str)
+        pk = kwargs.pop("pk")
```
@bradenmacdonald re the [slug / key addition](https://github.com/openedx/edx-platform/pull/35321#issuecomment-2323788286) -- should we use the `collection.key` or the full LibraryCollectionKey in these view URLs instead of the `collection.pk`?

bradenmacdonald replied:
I would just use the collection slug in the URL. Basically I think the only place we should use the opaque keys (`LibraryCollectionKey`) is for tagging since it requires an OpaqueKey, and everywhere else (if feasible) just use the library ID/key and the collection slug as separate parameters. Then we don't have to deal with opaquekeys anywhere else.

pomegranited committed:
temp: use temporary openedx-learning branch

pomegranited committed:
refactor: use collection.key as ID in search records

pomegranited committed:
refactor: use Collection.key as identifier in content_libraries

and refactors collection views to use DRF conventions.

pomegranited committed:
test: adds tests for events raised by collections api

pomegranited committed:
temp: use temporary openedx-events branch

rpenido commented on openedx/core/djangoapps/content_libraries/views.py:
```diff
@@ -136,12 +137,18 @@ def convert_exceptions(fn):
             return fn(*args, **kwargs)
+        except InvalidKeyError as exc:
+            log.exception(str(exc))
+            raise NotFound  # lint-amnesty, pylint: disable=raise-missing-from
```
Adding a new collection with the same key/library pair throws `django.db.utils.IntegrityError`. Maybe also add it here re-raising as `ValidationError`

rpenido commented on openedx/core/djangoapps/content_libraries/serializers.py:
```diff
@@ -245,3 +250,60 @@ class ContentLibraryBlockImportTaskCreateSerializer(serializers.Serializer):
+    """
+
+    title = serializers.CharField()
+    description = serializers.CharField()
```
As commented in the other PR, I think we need something like this to allow updating the `key`:
```suggestion
    description = serializers.CharField()
    key = serializers.CharField()
```

pomegranited replied:
Nope, addressed here https://github.com/openedx/openedx-learning/pull/223#discussion_r1746402764

rpenido reviewed (commented):
LGTM :+1:
Thank you for your work, @pomegranited!

Only a nit about updating the `key` and improving our error handling.

- [x] I tested this using the instructions from the PR
- [x] I read through the code
- [x] ~I checked for accessibility issues~
- [x] Includes documentation

pomegranited committed:
fix: catch IntegrityError in create_library_collection

and re-raise as api.LibraryCollectionAlreadyExists

pomegranited committed:
Merge remote-tracking branch 'origin/master' into yusuf-musleh/collections-crud-rest-api

pomegranited committed:
chore: updates openedx-events==9.14.0

and fixes event types documented in the hooks.

pomegranited commented on requirements/constraints.txt:
```diff
@@ -93,7 +93,7 @@ libsass==0.10.0
 
 # pinning this version to avoid updates while the library is being developed
-openedx-learning==0.11.2
+openedx-learning @ git+https://github.com/open-craft/openedx-learning/@jill/collection-key
```
TO DO: fix this dependency once https://github.com/openedx/openedx-learning/pull/223 merged & tagged

navinkarkera commented on openedx/core/djangoapps/content_libraries/views_collections.py:
```diff
@@ -0,0 +1,184 @@
+    """
+
+    serializer_class = ContentLibraryCollectionSerializer
+    lookup_field = 'key'
```
@pomegranited Since we plan to [make use](https://github.com/openedx/frontend-app-course-authoring/issues/1101#issuecomment-2300957646) of `object_key` i.e. `LibraryCollectionKey` in frontend urls, we should also convert this view to accept `LibraryCollectionKey` instead of `key` similar to [libraries](https://github.com/open-craft/edx-platform/blob/d312ff3e6aa1f02785f2cd2dd68d54df02359ad1/openedx/core/djangoapps/content_libraries/views.py#L290-L299). **Provided** we really want to include it, as we already have library key in the url in both frontend and backend.

I can do it in a follow up PR if we don't want to block this one.

ChrisChV replied:
Same, there are different places where `LibraryCollectionKey` can be used. I agree with creating a task for it and doing it when opaque-keys is [updated in edx-platform](https://tasks.opencraft.com/browse/FAL-3785?focusedCommentId=296651&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-296651)

bradenmacdonald replied:
For now I would prefer to not use the opaque keys in too many places, other than for tagging. We actually tried to avoid using the opaque key completely, except that it's required for tagging. As I mentioned in https://github.com/openedx/edx-platform/pull/35321#discussion_r1742724186 for the URLs and APIs we can generally just accept both a library ID and a `key`/slug (not the full opaque key, just the key part) as separate URL parameters. The python APIs can just use the collection integer IDs which is more efficient than huge string keys anyways.

navinkarkera replied:
@bradenmacdonald Yes, that is much better.

ChrisChV commented:
@pomegranited I created https://github.com/open-craft/edx-platform/pull/683 with some small changes. Those changes are used in https://github.com/openedx/frontend-app-course-authoring/pull/1259, I've seen that it's easier to add them in this PR. Let me know what you think

ChrisChV reviewed (approved):
@pomegranited Looks good :+1: I will wait for https://github.com/openedx/openedx-learning/pull/223 and for your review in https://github.com/open-craft/edx-platform/pull/683

pomegranited committed:
Merge branch 'master' into yusuf-musleh/collections-crud-rest-api

pomegranited committed:
chore: updates openedx-learning==0.11.4

ChrisChV committed:
refactor: Update collections crud rest api (#683)

* Update description as optional in ContentLibraryCollectionUpdateSerializer
* Create collection Rest API to auto-generate key

rpenido committed:
fix: assert collection doc have unique id

rpenido commented:
Hi @ChrisChV!
I found a small issue in the `id` for the document in the meilisearch index.
I updated it here: 1738447695c56069e13a9474a338e38f29d09c81

Our `key` field is not unique (we can have the same key for multiple libraries).
As this `id` is for meilisearch only and not exported or used elsewhere, I think it is safe to use the Collection `id` instead of the key.

ChrisChV commented:
@rpenido Looks good and works well :+1:

ChrisChV committed:
Merge branch 'master' into yusuf-musleh/collections-crud-rest-api

pomegranited commented:
@rpenido 
> As this id is for meilisearch only and not exported or used elsewhere, I think it is safe to use the Collection id instead of the key.

Good catch on the `id` -- but we still need the Collection.key to be present in that data, because that's required for accessing the REST APIs.

Will follow up with on this with [FAL-3787](https://tasks.opencraft.com/browse/FAL-3787) CC @ChrisChV
